File: Code/CPP/problem003/euler003.cpp
#include <iostream>
#include <cmath>
#include <memory>
#include "PrimeNumbers.h"

const unsigned long number = 600851475143L;

int main(int argc, char * argv[]) 
{
  /* constexpr unsigned long sqrtnum = ceil( sqrt( number ) ); */

  std::shared_ptr<PrimeNumbers> calculator(new PrimeNumbers());
  unsigned long startingPoint = ceil( sqrt(number) );
  startingPoint = calculator->nextPrime( startingPoint );
  std::cout << "Investigating : " << number << std::endl;
  std::cout << "Starting Point: " << startingPoint << std::endl;

  unsigned long value = startingPoint;

  while (value > 0) {
    if ( calculator->isPrime(value) && (0 == number % value) ) {
      std::cout << "Largest Prime Factor: " << value << std::endl;
      break;
    }
    --value;
  }

  return 0;
}

File: Code/R/problem003/euler003.R
#!/usr/bin/env Rscript

is.prime.tailrec <- function(n, m) { 
  if (m == 1) TRUE else { if (n %% m == 0) FALSE else is.prime.tailrec(n, m - 1) }
}

is.prime <- function(n) { is.prime.tailrec(n, n - 1) }

next.prime <- function(n) {
  n <- n + 1
  while ( !is.prime(n) ) {
    n <- n + 1
  }
  n
}

list.of.primes <- function(n) {
  answer <- c(1)
  num <- 2
  while (num < n) {
    if (is.prime(num)) answer <- append(answer, num)
    num <- num + 1
  }
  answer
}

query.number <- 13195
#query.number <- 60085
#query.number <- 600851475143

answer <- c(1)

starting.point <-  next.prime( ceiling( sqrt( query.number ) ) )
primes <- list.of.primes(starting.point)

index <- length(primes)
largest.prime.factor <- primes[index]

while ( 0 != query.number %% largest.prime.factor ) {
  index <- index - 1
  largest.prime.factor <- primes[index]
}

print(largest.prime.factor)



